lock-analyzer/src/main.rs

```diff
@@ -0,0 +1,215 @@
+                    Entry::Occupied(occupied) => {
+                        let rank = occupied.get();
+                        assert_eq!(rank.member_name, member_name);
+                        assert_eq!(rank.const_name, const_name);
+                    }
+                    Entry::Vacant(vacant) => {
+                        vacant.insert(Rank {
+                            member_name,
+                            const_name,
+                            acquisitions: BTreeMap::default(),
+                        });
+                    }
+                },
+                Action::Acquisition {
+                    older_rank,
+                    older_location,
+                    newer_rank,
+                    newer_location,
+                } => {
+                    let older_location = locations[&older_location].clone();
+                    let newer_location = locations[&newer_location].clone();
+                    ranks
+                        .get_mut(&older_rank)
+                        .unwrap()
+                        .acquisitions
+                        .entry(newer_rank)
+                        .or_default()
+                        .entry(older_location)
+                        .or_default()
+                        .insert(newer_location);
+                }
+            }
+        }
+    }
+
+    for older_rank in ranks.values() {
+        if older_rank.acquisitions.is_empty() {
+            // We'll print leaf locks separately, below.
+            continue;
+        }
+        println!(
+            "    rank {} {:?} followed by {{",
+            older_rank.const_name, older_rank.member_name
+        );
+        let mut first_newer = true;
+        for (newer_rank, locations) in &older_rank.acquisitions {
+            if !first_newer {
+                println!();
+            }
+            for (older_location, newer_locations) in locations {
+                if newer_locations.len() == 1 {
+                    for newer_loc in newer_locations {
+                        println!("        // holding {older_location} while locking {newer_loc}");
+                    }
+                } else {
+                    println!("        // holding {older_location} while locking:");
+                    for newer_loc in newer_locations {
+                        println!("        //     {newer_loc}");
+                    }
+                }
+            }
+            println!("        {},", ranks[newer_rank].const_name);
+            first_newer = false;
+        }
+        println!("    }};");
+        println!();
+    }
+
+    for older_rank in ranks.values() {
+        if !older_rank.acquisitions.is_empty() {
+            continue;
+        }
+
+        println!(
+            "    rank {} {:?} followed by {{ }};",
+            older_rank.const_name, older_rank.member_name
+        );
+    }
+
+    Ok(())
+}
+
+#[derive(Debug, serde::Deserialize)]
+enum Action {
```

Comment:
The `usize` vs. `u64` question is another reason to keep the two definitions of `Action` separate.

Reply:
Also I'm going to pretty hard veto "#[path]` anywhere, it throws RA for a total loop.

Reply:
@cwfitzgerald, @jimblandy: Ah, I hadn't been aware of that until I tried it myself. Yuck! Upstream tracking: https://github.com/rust-lang/rust-analyzer/issues/3898

If this were not an issue, I actually have working local changes that use the following to resolve @jimblandy's concerns (which I would have submitted as a follow-up PR):

* `&str` vs. `String` can be papered over with `Cow<'_, str>`.
* `usize` and `u64` can be separated out with a generic. We _could_ also just use `u64` everywhere, and widen pointers as necessary in `wgpu-core`.

…but I'm not willing to break `rust-analyzer` flow for something controversial. lezduit!

> If there are problems, we'll get errors from serde.

This is correct with this current code. I actually just noticed that there's one place where this might be wrong if we ever add an optional field and don't use `serde(deny_unknown_fields)` ([playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=6ab13214d2463b877d102f4a9ab6ed79)). This isn't blocking, since the case doesn't apply here, but I've pushed up the fix with e13c61405, since the fix is easy.